app/src/main/java/com/ahugenb/hra/calculator/CalculatorView.kt

```diff
@@ -0,0 +1,158 @@
+                    if (it.isSanitized() && it.smartToDouble().isValidPercent()) {
+                        abv.value = it
+                        viewModel.updateCalculation(
+                            abv.value.smartToDouble(),
+                            volume.value.smartToDouble(),
+                            drinks.value.smartToDouble(),
+                            mlChecked.value
+                        )
+                    }
+                },
+                label = { Text(text = "% ABV") }
+            )
+            OutlinedTextField(
+                maxLines = 1,
+                modifier = Modifier.weight(1f, true),
+                value = drinks.value,
+                keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Decimal),
+                onValueChange = {
+                    if (it.isSanitized() && it.smartToDouble().isValidDrinks()) {
+                        drinks.value = it
+                        viewModel.updateCalculation(
+                            abv.value.smartToDouble(),
+                            volume.value.smartToDouble(),
+                            drinks.value.smartToDouble(),
+                            mlChecked.value
+                        )
+                    }
+                },
+                label = { Text(text = "No. of Drinks") }
+            )
+        }
+        Text(text = stringResource(labelId))
+        Switch(
+            checked = mlChecked.value,
+            onCheckedChange = {
+                mlChecked.value = it
+
+                viewModel.updateCalculation(
+                    abv.value.smartToDouble(),
+                    volume.value.smartToDouble(),
+                    drinks.value.smartToDouble(),
+                    mlChecked.value
+                )
+            }
+        )
+
+        val units = viewModel.units.collectAsState().value
+        val pureEthanol = viewModel.pureEthanol.collectAsState().value
+        Text(
+            textAlign = TextAlign.Right,
+            text = stringResource(R.string.hra_units, units), style = MaterialTheme.typography.h4,
+            modifier = Modifier
+                .padding(horizontal = 8.dp)
+                .fillMaxWidth(),
+        )
+        Text(
+            textAlign = TextAlign.Right,
+            text = stringResource(ethanolId, pureEthanol),
+            style = MaterialTheme.typography.h4,
+            modifier = Modifier
+                .padding(horizontal = 8.dp)
+                .fillMaxWidth()
+        )
+    }
+}
+
+/*
+* sanitized input:
+*  -is less than 10 digits in length
+*  -may not contain more than one decimal
+*  -may only contain digits from validInput
+*/
+private fun String.isSanitized(): Boolean {
+    val validInput = "1234567890."
+
+    return (this.isEmpty() || this.length < 10 && this.count { ch -> ch == '.' } < 2
+            && this.all { ch -> validInput.contains(ch) })
+}
+
+private fun Double.isValidPercent(): Boolean = this in 0.0..100.0
```

Comment:
Not sure if you're going to ever reuse this code. But you could move these elsewhere like a`Double` utility functions.